doc/architecture/decisions/0016-web-container-interface.md

```diff
@@ -0,0 +1,96 @@
+# 16. Container Interface for web UI clients
+
+Date: 2018-09-14
+
+## Status
+
+Accepted
+
+## Context
+* We have an implementation of a Holochain container that can load QML based UIs: https://github.com/holochain/holosqape
+* We also want to support HTML/JS/CSS based UIs
+* We need to enable pure Web clients to connect to a HoloPort for the Holo use-case
+* With [ADR 13](https://github.com/holochain/holochain-rust/blob/develop/doc/architecture/decisions/0013-signals-listeners-model-and-api.md)
+  we defined *app signals*, i.e. events that a Holochain app can push - we need to make sure UIs can receive these as
+  pushed notifications from the Holochain node 
+    
+
+## Requirements
+It seems we need the advantages of both worlds:
+* having a rich context in which apps run so we can do bridging and other sophisticated
+  app composability features (the need for a container/composer)
+* we need clients that are running any kind of UI (or connecting with IoT devices)
+  to be able to connect with Holochain apps easily
+### Mandatory
+- Any full-node Holochain instance should be online most of the time (i.e. not tied to the browser process)
+  to be available to the DHT, which makes it unattractive to run the whole instance in a web browser.
+
+
+### Desirable
+- Bi-directional communication (pub/sub)
+    From the UI client side. It should be possible to subscribe to certain events such as when a node receives new data, receives a message etc.
+- Exposed through an object abstraction in the browser runtime (like web3)
+- Can also be used to connect to a remote node for compatability with Holo
+
+## Considered Solutions
+
+### 1. Daemon container that accepts function calls via HTTP requests
+This is the solution implemented in proto. 
+- A single daemon container exposes end-points for calling functions namespaced by app and by zome
+- Main drawback is that state changes need to be polled by UI - no push notifications
+
+### 2. Daemon container that exposes WebSockets for bi-directional communication
```

Comment:
there's also server sent events (SSE) and long polling to consider.

server sent events: 

- pushes from the server. 
- natively supported by everything except IE/edge.
- polyfill available for IE/edge
- load balancer/CDN support (cloudfront chokes on websockets)
- native reconnection handling

long polling:

- different to a dumb poll loop
- client polls, server sends data immediately if available, otherwise holds open connection until it is available
- client immediately reconnects after receiving data

more info on websockets vs. SSE vs. long poll:

https://blog.stanko.io/do-you-really-need-websockets-343aed40aa9b